### ISSUE TITLE
Support for smaller in-memory HT with outer join conditions

### ISSUE THREAD

**Issue by jtommaney-alibaba** [cwida/duckdb#1020]
Executed join with 10k row table x 1 billion row table under left/right outer join options.

With existing implementation when the outer join logic requires the large table to be in-memory there is:
1)  a significant drop in performance ( 82 seconds vs  0.75 seconds)
2) a significant rise in memory utilization ( 20gb vs > 1mb) 

Recommend enabling logic to allow for outer join table to sit in memory. 
Pseudo-logic:
  scan/probe from larger table will flag hit on entries in in-memory HT.
  finalize operation will scan in-memory HT identifying entries not flagged and insert tuples 

```
D select count(*) from t_10k d join t_1b f on ( d.pkey=f.c_10k);
┌────────────┐
│  count()   │
├────────────┤
│ 1000000000 │
└────────────┘
Run Time: real 0.756 user 13.680745 sys 10.041263
D select count(*) from t_10k d left outer join t_1b f on ( d.pkey=f.c_10k);
┌────────────┐
│  count()   │
├────────────┤
│ 1000000000 │
└────────────┘
Run Time: real 82.144 user 161.324755 sys 32.406965
D select count(*) from t_10k d right outer join t_1b f on ( d.pkey=f.c_10k);
┌────────────┐
│  count()   │
├────────────┤
│ 1000000000 │
└────────────┘
Run Time: real 0.747 user 15.778193 sys 7.515481
```

```
CREATE TABLE t_100m ( pkey integer, c0 char(2), c_1k integer, c_10k integer, c_100k integer, c_1m integer, c_10m integer, c_100m integer );

insert into t_100m
select i, concat('A',mod(i,2)), (random()* 1000)::int, (random()* 10000)::int, (random()* 100000)::int, 
         (random()* 1000000)::int, (random()* 10000000)::int, (random()* 100000000)::int  FROM range(0,100000000) tbl(i) ;

CREATE TABLE t_1b ( pkey integer, c0 char(2), c_1k integer, c_10k integer, c_100k integer, c_1m integer, c_10m integer, c_100m integer );

insert into t_1b
select i, concat('A',mod(i,2)), (random()* 1000)::int, (random()* 10000)::int, (random()* 100000)::int, 
         (random()* 1000000)::int, (random()* 10000000)::int, (random()* 100000000)::int  FROM range(0,1000000000) tbl(i) ;


select count(*) from t_10k d join t_1b f on ( d.pkey=f.c_10k);
select count(*) from t_10k d left outer join t_1b f on ( d.pkey=f.c_10k);
select count(*) from t_10k d right outer join t_1b f on ( d.pkey=f.c_10k);



```

**Comment by jtommaney-alibaba** [comment]
TPCH Q13 also is an example:

select 
 c_count, count(*) as custdist
from (select c_custkey, count(o_orderkey)
      from customer left outer join orders on c_custkey = o_custkey
      and o_comment not like '%special%%requests%'
      group by c_custkey) as c_orders 
(c_custkey, c_count)
group by c_count order by custdist desc, c_count desc;

**Comment by Mytherin** [comment]
Thanks for the issue! Good idea indeed. This should not be too difficult to implement since we already implement the `FULL OUTER JOIN`, which has a similar implementation on the right-hand side.

**Comment by jtommaney-alibaba** [comment]
The full outer join performance implementation in duckdb is very fast.  

join                                       0.114
full outer+ restrict             0.950      8.3x slower 
right outer join                 14.600    128x slower

Optimizing this piece has two significant benefits.
   1)  significant speedup and reduced memory when small side is outer.  (close to 128x ?  )   
   2)  Join order no longer has a dependency on left/right/outer conditions.  
                    Join, left outer, right outer, full outer all should have the same optimal plan.  

```
drop table d;
create temporary table d as select range k, mod(range, 3) + 1 g, 42 p from range(30000000);
drop table e;
create temporary table e as select range k, mod(range, 3) + 2 g, 42 p from range(9);

select d.g, e.g, count(*) from d join e using (g) group by 1,2 order by 1,2  ;
select d.g, e.g, count(*) from d left outer join e using (g) group by 1,2 order by 1,2  ;
select d.g, e.g, count(*) from d right outer join e using (g) group by 1,2 order by 1,2  ;
select d.g, e.g, count(*) from d full outer join e using (g) group by 1,2 order by 1,2  ;
select d.g, e.g, count(*) from d full outer join e using (g) 
where e.g is not null group by 1,2 order by 1,2  ;

D select d.g, e.g, count(*) from d right outer join e using (g) group by 1,2 order by 1,2  ;
┌───┬───┬──────────┐
│ g │ g │ count()  │
├───┼───┼──────────┤
│   │ 4 │ 3        │
│ 2 │ 2 │ 30000000 │
│ 3 │ 3 │ 30000000 │
└───┴───┴──────────┘
Run Time: real 14.600 user 29.269942 sys 1.211979
D select d.g, e.g, count(*) from d full outer join e using (g) group by 1,2 order by 1,2  ;
┌───┬───┬──────────┐
│ g │ g │ count()  │
├───┼───┼──────────┤
│   │ 4 │ 3        │
│ 1 │   │ 10000000 │
│ 2 │ 2 │ 30000000 │
│ 3 │ 3 │ 30000000 │
└───┴───┴──────────┘
Run Time: real 0.941 user 14.887059 sys 0.036436
D 
D 
D select d.g, e.g, count(*) from d full outer join e using (g) 
   ...> where e.g is not null group by 1,2 order by 1,2  ;
┌───┬───┬──────────┐
│ g │ g │ count()  │
├───┼───┼──────────┤
│   │ 4 │ 3        │
│ 2 │ 2 │ 30000000 │
│ 3 │ 3 │ 30000000 │
└───┴───┴──────────┘
Run Time: real 0.950 user 15.011023 sys 0.061768
```

**Comment by Mytherin** [comment]
This should be fixed and implemented now in #1097. Thanks again for the very detailed report! :)
